drv/mb85rsxx-fram/src/lib.rs

```diff
@@ -0,0 +1,428 @@
+            .lock_auto(CsState::Asserted)
+            .map_err(|_| SpiError::TaskRestarted)?;
+        let mut buf = [0; 3];
+        spi.exchange(&[Opcode::ReadId as u8, 0, 0], &mut buf)?;
+        let [_, mfg_id, cont] = buf;
+
+        let product_id = if cont == CONTINUE {
+            let mut buf = [0; 2];
+            spi.read(&mut buf)?;
+            u16::from_be_bytes(buf)
+        } else {
+            // no continuation code --- use the bytes we just read.
+            // AFAICT, the Fujitsu FRAM chips don't do this, but other
+            // manufacturers' FRAM chips that use the same protocol do? Would
+            // have to read some more datasheets to be sure.
+            u16::from_be_bytes([mfg_id, cont])
+        };
+
+        Ok(Self { mfg_id, product_id })
+    }
+}
+
+impl<S: SpiServer, const ID: u16> Drop for WritableFram<'_, S, { ID }> {
+    fn drop(&mut self) {
+        // Put the FRAM back the way we found it.
+        let _ = self.0.do_write_disable();
+    }
+}
+
+impl From<SpiError> for FramInitError {
+    fn from(value: SpiError) -> Self {
+        FramInitError::SpiError(value)
+    }
+}
+
+/// Product IDs for various Fujitsu FRAM devices.
+///
+/// This driver primarily uses the product ID to determine whether the size of
+/// the FRAM chip matches the expected size. Other features of the FRAM chip can
+/// also be determined from the product ID, as well; however, for the purposes
+/// of this driver, we basically only care about the size of the device, because
+/// we don't use some of the fancier features, such as bank protection or the
+/// HOLD_L and WP_L pins.
+pub mod product_id {
+
+    /// 2kb Fujitsu FRAM
+    pub const MB85RS16: u16 = 0x0101;
+    /// 8kb Fujitsu FRAM
+    pub const MB85RS64V: u16 = 0x0302;
+    /// 8kb Fujitsu FRAM
+    pub const MB85RS64T: u16 = 0x2303;
+    /// 32kb Fujitsu FRAM
+    pub const MB85RS256TY: u16 = 0x2503;
+    /// 128kb Fujitsu FRAM
+    pub const MB85RS1MT: u16 = 0x2703;
+    /// 256kb Fujitsu FRAM
+    pub const MB85RS2MTA: u16 = 0x4803;
+    /// 512kb Fujitsu FRAM
+    pub const MB85RS4MT: u16 = 0x4903;
+
+    /// Returns the size in bytes of the FRAM chip, based on its product ID.
+    pub(super) const fn size(product_id: u16) -> usize {
+        // The first 5 bits of the product ID give the density of the FRAM chip
+        // in multiples of 2KiB.
+        //
+        // For example, the 2kb MB85RS16 has the product ID 0x0101, so:
+        //  0x01 & 0b0001_1111 = 1
+        //  2^1 = 2
+        //  2 * 1024 = 2048 bytes
+        //
+        // Or, for the 8kb MB85RS64V and MB85RS64T, which have product IDs
+        // 0x0302 and 0x2303, respectively:
+        //  0x03 & 0b0001_1111 = 3
+        //  0x23 & 0b0001_1111 = 3
+        //  2^3 = 8
+        //  8 * 1024 = 8192 bytes
+        const MASK: u8 = 0b0001_1111;
+        let [hi, _] = u16::to_be_bytes(product_id);
+        let density = hi & MASK;
+        2usize.pow(density as u32) * super::KIB
```

Comment:
Weird personal opinion: even though `pow` is more mathematically obvious, I typically see conversion from log2 size → actual size written as a shift, e.g. `(1usize << density as u32)`